gef-ui/server/controllers/unissued-facilities/facility-end-date/index.ts

```diff
@@ -0,0 +1,217 @@
+  const {
+    params: { dealId, facilityId },
+    session: { userToken },
+  } = req;
+
+  const { details: facility } = (await api.getFacility({ facilityId, userToken })) as { details: Record<string, unknown> };
+  const deal = (await api.getApplication({ dealId, userToken })) as Record<string, unknown> & { version?: number };
+
+  if (!isFacilityEndDateEnabledOnGefVersion(parseDealVersion(deal.version)) || !facility.isUsingFacilityEndDate) {
+    return null;
+  }
+
+  const facilityEndDateViewModel: FacilityEndDateViewModel = {
+    dealId,
+    facilityId,
+    previousPage,
+  };
+
+  if ('facilityEndDate' in facility && typeof facility.facilityEndDate === 'string') {
+    const facilityEndDate = parseISO(facility.facilityEndDate);
+    facilityEndDateViewModel.facilityEndDate = {
+      day: String(getDate(facilityEndDate)),
+      month: String(getMonth(facilityEndDate) + 1),
+      year: String(getYear(facilityEndDate)),
+    };
+  }
+
+  return facilityEndDateViewModel;
+};
+
+export const getFacilityEndDateFromUnissuedFacilitiesPage = async (req: GetFacilityEndDateRequest, res: Response) => {
+  const {
+    params: { dealId, facilityId },
+  } = req;
+
+  const previousPage = `/gef/application-details/${dealId}/unissued-facilities/${facilityId}/about`;
+
+  try {
+    const facilityEndDateViewModel = await getFacilityEndDateViewModel(req, previousPage);
+
+    if (!facilityEndDateViewModel) {
+      return res.redirect(previousPage);
+    }
+
+    return res.render('partials/facility-end-date.njk', facilityEndDateViewModel);
+  } catch (error) {
+    console.error(error);
+    return res.render('partials/problem-with-service.njk');
+  }
+};
+
+export const getFacilityEndDateFromApplicationPreviewPage = async (req: GetFacilityEndDateRequest, res: Response) => {
+  const {
+    params: { dealId, facilityId },
+  } = req;
+
+  const previousPage = `/gef/application-details/${dealId}/unissued-facilities/${facilityId}/change`;
+
+  try {
+    const facilityEndDateViewModel = await getFacilityEndDateViewModel(req, previousPage);
+
+    if (!facilityEndDateViewModel) {
+      return res.redirect(previousPage);
+    }
+    return res.render('partials/facility-end-date.njk', facilityEndDateViewModel);
+  } catch (error) {
+    console.error(error);
+    return res.render('partials/problem-with-service.njk');
+  }
+};
+
+const getCoverStartDateOrStartOfToday = (facility: Record<string, unknown>): Date => {
+  if (typeof facility.coverStartDate === 'string') {
+    return startOfDay(parseISO(facility.coverStartDate));
+  }
+
+  if (!facility.coverStartDate) {
+    return startOfDay(new Date());
+  }
+
+  throw new Error('Invalid coverStartDate');
+};
+
+export const validateAndReturnErrorsOrUpdateFacilityEndDate = async (req: PostFacilityEndDateRequest): Promise<ValidationError[] | null> => {
+  const {
```

Comment:
out of interest why do we prefer returning null as apposed to void

Reply:
We use this [linting rule](https://eslint.org/docs/latest/rules/consistent-return) which gets sad if you have a combination of `return` and `return { foo: 'bar' }` in the same function. I've managed to make the return type `Promise<ValidationError[] | void>` by writing `return undefined` but i think i prefer `return null` - what are your thoughts?

Tbh I'm not 100% happy with this function but so if you have better ideas more than happy to implement them